### PR TITLE
[WIP] Display function creation error in UI

### PR DESCRIPTION
**Pull request by nullseed**
## Description
- If `POST /system/functions` errors then a toast is displayed to the user with the error message and the dialog remains open.
- If `POST /system/functions` succeeds then a toast is displayed to the user with the message "Function created" and the dialog closes.

## Motivation and Context
Related to #195 

## How Has This Been Tested?

Manually testing the UI

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: gateway/assets/script/bootstrap.js [hunk 1]
```diff
@@ -130,20 +129,21 @@ app.controller("home", ['$scope', '$log', '$http', '$location', '$timeout', '$md
                 url: "/system/functions",
                 data: $scope.item,
                 method: "POST",
-                headers: { "Content-Type": "application/json"},
-                responseType: "json"
+                headers: { "Content-Type": "application/json" },
+                responseType: "text"
```
Comment (alexellis) [inline gateway/assets/script/bootstrap.js]: +                responseType: "text" ?

**Comment by alexellis** [comment]
Nice to see this. What about a little gif to show it in action?

File: gateway/assets/script/bootstrap.js (hunk 1 above)
Comment (nullseed) [reply gateway/assets/script/bootstrap.js]: `w.Write([]byte("Deployment error: " + err.Error()))` writes plain text and not json, if responseType is set to json then angular attempts to parse it and the data ends up being null

**Comment by nullseed** [comment]
create-function-ui

**Comment by alexellis** [comment]
Thank you, very clear and much better.

I think the toast is OK for the errors - but I was also wondering if we could relate it to the pop-up somehow - like putting text in the dialog too.

**Comment by nullseed** [comment]
Yeah it would definitely be good to have the validation errors next to the fields. This adds a little complexity on the gateway side because we'd have to format error messages in a way that relates errors to fields e.g. json.

**Comment by nullseed** [comment]
Successful function creation still displays a toast, errors are now at the bottom of the form.

**Comment by alexellis** [comment]
We removed the link to `<link rel="stylesheet" href="style/angular-material.min.css">`?

I would have thought we need this and if it's missing we may need to add it again?

**Comment by nullseed** [comment]
From looking at the [history](https://github.com/openfaas/faas/commits/master/gateway/assets/style) of the `assets/style` folder it looks like `angular-material.min.css` was never added. The link in the markup was added [a while back](https://github.com/openfaas/faas/blame/master/gateway/assets/index.html#L7). The stylesheet is being used from two lines above it [here](https://github.com/openfaas/faas/blob/master/gateway/assets/index.html#L5). Removing this line removed the 404 error i was getting when the browser tried to load the stylesheet

File: gateway/assets/newfunction.html [hunk 2]
```diff
@@ -1,40 +1,42 @@
-<md-dialog  aria-label="List dialog" layout="column" flex="70"> 
+<md-dialog aria-label="List dialog" layout="column" flex="70">
     <md-dialog-content class="md-padding">
         <label>Define a function:</label>
         <form name="userForm">
             <div layout-gt-xs="row">
                 <md-input-container class="md-block" flex-gt-sm>
-                <label>Image:</label>
-                <input name="dockerImage" ng-model="item.image" required md-maxlength="200" minlength="4">
```
Comment (alexellis) [inline gateway/assets/newfunction.html]: We also took off `required`?

File: gateway/assets/script/bootstrap.js (hunk 1 above)
Comment (alexellis) [reply gateway/assets/script/bootstrap.js]: OK

File: gateway/assets/script/bootstrap.js [hunk 3]
```diff
@@ -130,20 +129,19 @@ app.controller("home", ['$scope', '$log', '$http', '$location', '$timeout', '$md
...
             };
 
             $http(options)
                 .then(function(response) {
-                    $scope.invocationResponse = response.data;
-                    $scope.invocationStatus = response.status;
+                    item.image = item.service = item.envProcess = item.network = "";
```
Comment (alexellis) [inline gateway/assets/script/bootstrap.js]: I would separate this out onto separate lines for readability.

File: gateway/assets/newfunction.html (hunk 2 above)
Comment (alexellis) [reply gateway/assets/newfunction.html]: I carved out a few minutes to test the PR on my laptop - seems like removing the `required` field may have introduced regression in the UI minimum length / required field status.

I can create a service with no name.. that's not expected behaviour.

Comment (nullseed) [reply gateway/assets/newfunction.html]: I've added the attribute back but i'm still able to submit an empty form. When i click `Create` in the new function dialog I get the error `Deployment error: Error response from daemon: rpc error: code = 3 desc = ContainerSpec: image reference must be provided` which shows that a trip to the server is happening.

**Comment by nullseed** [comment]
Oh Derek

**Comment by alexellis** [comment]
Yes so we've mixed a bit of Slack + GitHub - I tested and the `required` attribute didn't resolve the issue. I mentioned that it needs further investigation to trigger the angular validation, or only allow the click if it's valid in the model/state.

**Comment by alexellis** [comment]
Could do with squashing the commits down too 👍 

```
$ git reset HEAD~N     # N is the number of commits you made
$ git commit -s            # sign-off all changes at once
# git push origin BRANCH_NAME
```

**Comment by nullseed** [comment]
Squashed commits into one and signed off - see above.

**Comment by alexellis** [comment]
https://github.com/openfaas/faas/issues/252

**Pull request merged**